### PR TITLE
Adding possibility to 'cast' or copy to `xt::xarray` etc

### PR DESCRIPTION
**Pull request by tdegeus**

### CODE REVIEW

**Review by JohanMabille — COMMENTED** [review]
I'm not sure it would make sense to implement the cast operator for `strided_views` (which is used for creating a view on an existing expression).

It would make sense to implement it for the `xarray_adaptor`, though. It would be a non-owning adpator on the underlying buffer.

File: include/xtensor-python/xtensor_type_caster_base.hpp [hunk 1]
```diff
@@ -106,6 +183,30 @@ namespace pybind11
...
+                auto buf = xtensor_check_buffer<Type>::get(src);
+
+                if (!buf) {
+                    return false;
+                }
+
+                std::vector<size_t> shape(buf.ndim());
+                std::copy(buf.shape(), buf.shape() + buf.ndim(), shape.begin());
+                value = Type(shape);
+                std::copy(buf.data(), buf.data() + buf.size(), value.begin());
```
[the remark this answers is not in the document]
Comment (tdegeus) [reply include/xtensor-python/xtensor_type_caster_base.hpp:205]: Ah, I did not know about the pystride adaptor. So for the moment I use pybind11's way to ensure that data is contiguous, by using https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html#arrays. Personally I have no big feelings about one or the other, but at the same time I am not entirely sure about the difference in efficiency.

File: include/xtensor-python/xtensor_type_caster_base.hpp [hunk 2]
```diff
@@ -23,6 +23,87 @@ namespace pybind11
 {
     namespace detail
     {
+        template <typename T, xt::layout_type L>
+        struct xtensor_get_buffer
+        {
+            template <typename H>
+            static auto get(H src)
```
[the remark this answers is not in the document]
Comment (tdegeus) [reply include/xtensor-python/xtensor_type_caster_base.hpp:30]: You are right, made the change.

File: include/xtensor-python/xtensor_type_caster_base.hpp [hunk 3]
```diff
@@ -23,6 +23,87 @@ namespace pybind11
...
+            {
+                return array_t<T, array::f_style>::ensure(src);
+            }
+        };
+
+        template <class T>
+        struct xtensor_check_buffer
+        {
+        };
```
[the remark this answers is not in the document]
Comment (tdegeus) [reply include/xtensor-python/xtensor_type_caster_base.hpp:49]: I thought so too, but my compiler complained:
```cpp
xtensor_type_caster_base.hpp:50:16: error: no template named 'xtensor_check_buffer'; did you mean 'xtensor_get_buffer'?
        struct xtensor_check_buffer<xt::xarray<T, L>>
               ^~~~~~~~~~~~~~~~~~~~
               xtensor_get_buffer
```

**Comment by tdegeus** [comment]
Essentially this should be good to go. However, I had to introduce on additional check to check on rank for `xtensor`. In particular, I want to do this:
```cpp
static auto get(pybind11::handle src)
{
    auto buf = pybind11::array_t<T, pybind11::array::c_style | pybind11::array::forcecast>::ensure(src);
    if (buf.ndim() != 2) {
        return false;
    }
    return buf;
}
```
however my compiler gets (rightfully) confused:
```
error: 'auto' in return type deduced as 'pybind11::array_t<int, 17>' here but deduced as 'bool' in earlier return statement
                return buf;
                ^
```
So I worked around with the extra check. I would like to get rid of it. But I don't know what I can return instead of `false` that allows me to later do
```cpp
if (!buf) { 
    return false;
}
```

File: include/xtensor-python/xtensor_type_caster_base.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (JohanMabille) [reply include/xtensor-python/xtensor_type_caster_base.hpp]: Ah indeed `c_style` and `f_style` involve contiguity, I never paid attention to that, thanks for the pointer!

File: include/xtensor-python/xtensor_type_caster_base.hpp [hunk 4]
```diff
@@ -23,6 +23,97 @@ namespace pybind11
...
+
+        template <class T, xt::layout_type L>
+        struct xtensor_check_buffer<xt::xarray<T, L>>
+        {
+            static auto get(handle src)
+            {
+                return xtensor_get_buffer<T, L>::get(src);
+            }
+        };
```
Comment (JohanMabille) [inline include/xtensor-python/xtensor_type_caster_base.hpp:56]: Nitpicking: in line with the previous renaming proposal:
`xtensor_chek_buffer` => `pybind_array_getter`
`get` => `run`

File: include/xtensor-python/xtensor_type_caster_base.hpp [hunk 5]
```diff
@@ -23,6 +23,97 @@ namespace pybind11
...
+
+        template <class T>
+        struct xtensor_verify
+        {
+            template <class B>
+            static bool get(const B& buf)
+            {
+                return true;
+            }
+        };
```
Comment (JohanMabille) [inline include/xtensor-python/xtensor_type_caster_base.hpp:104]: Nitpicking:
`xtensor_verify` => `pybind_array_dim_checker`
`get` => `run`

File: include/xtensor-python/xtensor_type_caster_base.hpp [hunk 6]
```diff
@@ -106,6 +193,33 @@ namespace pybind11
 
         public:
 
+            PYBIND11_TYPE_CASTER(Type, _("numpy.ndarray[") + npy_format_descriptor<typename Type::value_type>::name + _("]"));
+
+            bool load(handle src, bool convert)
+            {
+                using T = typename Type::value_type;
+
+                if (!convert && !array_t<T>::check_(src)) {
```
Comment (JohanMabille) [inline include/xtensor-python/xtensor_type_caster_base.hpp:202]: Nitpicking: opening curly brace goes on a new line.

**Review by JohanMabille — COMMENTED** [review]
Looks good to me once the conflicts are solved. I've suggested some renaming that I find more explicit.

**Comment by tdegeus** [comment]
Thanks @JohanMabille ! Good to go.

**Pull request merged**